File: src/services/inlayHints.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { createChecker } from "../compiler/checker.ts";
import { DEFAULT_INLAY_HINTS, getInlayHints, type InlayHintsSettings } from "./inlayHints.ts";
import { loadJdkStub } from "../compiler/jdkStub.ts";
import { createProgram } from "../compiler/program.ts";
import { type Uri } from "../workspace.ts";

function hints(source: string, settings?: InlayHintsSettings) {
  const program = createProgram();
  loadJdkStub(program);
  program.setOpenDocument("file:///T.java" as Uri, source, 1);
  const checker = createChecker(program);
  const sourceFile = program.getSourceFile("file:///T.java" as Uri)!;
  return getInlayHints(checker, sourceFile, 0, source.length, settings).map(h => ({
    label: h.label,
    kind: h.kind,
    // The token the hint is attached to (start of an argument / end of a name).
    at: source.slice(h.offset, h.offset + 6),
  }));
}

test("parameter hints appear for literal and expression arguments only", () => {
  const out = hints(
    [
      "class C {",
      "  static int clamp(int value, int low, int high) { return value; }",
      "  void m(int x) {",
      "    int limit = 9;",
      "    clamp(x, 1 + 2, limit);", // x: plain var (skip), 1+2: expr (hint), limit: plain var (skip)
      "    clamp(5, x, compute());", // 5: literal (hint), x: skip, call: hint
      "  }",
      "  int compute() { return 0; }",
      "}",
    ].join("\n"),
  );
  const params = out.filter(h => h.kind === "parameter");
  expect(params.map(h => h.label)).toEqual(["low:", "value:", "high:"]);
  expect(params[0]!.at.startsWith("1 + 2")).toBe(true);
  expect(params[1]!.at.startsWith("5")).toBe(true);
  expect(params[2]!.at.startsWith("comput")).toBe(true);
});

test("varargs arguments hint only the first of the tail", () => {
  const out = hints(
    [
      "class C {",
      "  static int sum(int... xs) { return 0; }",
      "  void m() { sum(1, 2, 3); }",
      "}",
    ].join("\n"),
  );
  expect(out.filter(h => h.kind === "parameter").map(h => h.label)).toEqual(["...xs:"]);
});

test("var declarations and for-each get inferred type hints", () => {
  const out = hints(
    [
      "import java.util.List;",
      "class C {",
      "  void m(List<String> xs) {",
      '    var s = "hi";',
      "    var n = 1 + 2;",
      "    for (var item : xs) { use(item); }",
      "  }",
      "  void use(String s) {}",
      "}",
    ].join("\n"),
  );
  const types = out.filter(h => h.kind === "type");
  expect(types.map(h => h.label)).toEqual([": String", ": int", ": String"]);
});

test("settings disable each hint family independently", () => {
  const source = [
    "class C {",
    "  static int twice(int value) { return value * 2; }",
    "  void m() { var n = twice(21); }",
    "}",
  ].join("\n");
  expect(
    hints(source)
      .map(h => h.kind)
      .sort(),
  ).toEqual(["parameter", "type"]);
  expect(hints(source, { ...DEFAULT_INLAY_HINTS, parameterNames: false }).map(h => h.kind)).toEqual(
    ["type"],
  );
  expect(hints(source, { ...DEFAULT_INLAY_HINTS, varTypes: false }).map(h => h.kind)).toEqual([
    "parameter",
  ]);
});

File: togo/internal/services/inlay_hints_test.go
package services

import (
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/compiler"
)

// Port of src/services/inlayHints.test.ts.

type inlayHint struct {
	label, kind, at string
}

func inlayHints(source string, settings InlayHintsSettings) []inlayHint {
	program := compiler.NewProgram()
	compiler.LoadJdkStub(program)
	program.SetOpenDocument("file:///T.java", source, 1)
	checker := compiler.NewChecker(program)
	sourceFile := program.GetSourceFile("file:///T.java")
	var out []inlayHint
	for _, h := range GetInlayHints(checker, sourceFile, 0, len(source), settings) {
		end := h.Offset + 6
		if end > len(source) {
			end = len(source)
		}
		out = append(out, inlayHint{label: h.Label, kind: h.Kind, at: source[h.Offset:end]})
	}
	return out
}

func hintsOfKind(hints []inlayHint, kind string) []string {
	var out []string
	for _, h := range hints {
		if h.kind == kind {
			out = append(out, h.label)
		}
	}
	return out
}

func TestParameterHintsLiteralAndExpression(t *testing.T) {
	out := inlayHints(strings.Join([]string{
		"class C {",
		"  static int clamp(int value, int low, int high) { return value; }",
		"  void m(int x) {",
		"    int limit = 9;",
		"    clamp(x, 1 + 2, limit);",
		"    clamp(5, x, compute());",
		"  }",
		"  int compute() { return 0; }",
		"}",
	}, "\n"), DefaultInlayHints)
	params := hintsOfKind(out, "parameter")
	want := []string{"low:", "value:", "high:"}
	if len(params) != 3 || params[0] != want[0] || params[1] != want[1] || params[2] != want[2] {
		t.Errorf("parameter hints = %v, want %v", params, want)
	}
	var paramEntries []inlayHint
	for _, h := range out {
		if h.kind == "parameter" {
			paramEntries = append(paramEntries, h)
		}
	}
	if !strings.HasPrefix(paramEntries[0].at, "1 + 2") || !strings.HasPrefix(paramEntries[1].at, "5") || !strings.HasPrefix(paramEntries[2].at, "comput") {
		t.Errorf("attachment points wrong: %v", paramEntries)
	}
}

func TestVarargsHintFirstOfTail(t *testing.T) {
	out := inlayHints(strings.Join([]string{
		"class C {",
		"  static int sum(int... xs) { return 0; }",
		"  void m() { sum(1, 2, 3); }",
		"}",
	}, "\n"), DefaultInlayHints)
	params := hintsOfKind(out, "parameter")
	if len(params) != 1 || params[0] != "...xs:" {
		t.Errorf("varargs hints = %v, want [...xs:]", params)
	}
}

func TestVarAndForEachTypeHints(t *testing.T) {
	out := inlayHints(strings.Join([]string{
		"import java.util.List;",
		"class C {",
		"  void m(List<String> xs) {",
		"    var s = \"hi\";",
		"    var n = 1 + 2;",
		"    for (var item : xs) { use(item); }",
		"  }",
		"  void use(String s) {}",
		"}",
	}, "\n"), DefaultInlayHints)
	types := hintsOfKind(out, "type")
	want := []string{": String", ": int", ": String"}
	if len(types) != 3 || types[0] != want[0] || types[1] != want[1] || types[2] != want[2] {
		t.Errorf("type hints = %v, want %v", types, want)
	}
}

func TestSettingsDisableFamilies(t *testing.T) {
	source := strings.Join([]string{
		"class C {",
		"  static int twice(int value) { return value * 2; }",
		"  void m() { var n = twice(21); }",
		"}",
	}, "\n")
	all := inlayHints(source, DefaultInlayHints)
	kinds := map[string]bool{}
	for _, h := range all {
		kinds[h.kind] = true
	}
	if !kinds["parameter"] || !kinds["type"] {
		t.Errorf("default should have both families, got %v", kinds)
	}
	noParam := inlayHints(source, InlayHintsSettings{ParameterNames: false, VarTypes: true})
	if k := hintsOfKind(noParam, "parameter"); len(k) != 0 {
		t.Errorf("parameterNames=false should suppress parameter hints, got %v", k)
	}
	noType := inlayHints(source, InlayHintsSettings{ParameterNames: true, VarTypes: false})
	if k := hintsOfKind(noType, "type"); len(k) != 0 {
		t.Errorf("varTypes=false should suppress type hints, got %v", k)
	}
}
